Issue with specials caracters 'FF' (FORM FEED), 'CR' (CARIAGE RETURN) in shell scripts

Hi there,

I have an issue with the sonar-shellcheck plugin.

I have some specials caracters 'FF' (FORM FEED), 'CR' (CARIAGE RETURN) in shell scripts, and it doesn't work with the sonar-shellcheck plugin :
```
ERROR: Error during SonarScanner execution
java.lang.IllegalArgumentException: Unable to highlight file [...]
        at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:97)
        at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:37)
        at com.github.sbaudoin.sonar.plugins.shellcheck.highlighting.HighlightingData.highlight(HighlightingData.java:58)
        at com.github.sbaudoin.sonar.plugins.shellcheck.rules.ShellCheckSensor.saveSyntaxHighlighting(ShellCheckSensor.java:298)
        at com.github.sbaudoin.sonar.plugins.shellcheck.rules.ShellCheckSensor.execute(ShellCheckSensor.java:139)
        at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:48)
        at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:85)
        at org.sonar.scanner.sensor.ModuleSensorsExecutor.lambda$execute$1(ModuleSensorsExecutor.java:59)
        at org.sonar.scanner.sensor.ModuleSensorsExecutor.withModuleStrategy(ModuleSensorsExecutor.java:77)
        at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:59)
        at org.sonar.scanner.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:82)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
        at org.sonar.scanner.scan.ProjectScanContainer.scan(ProjectScanContainer.java:386)
        at org.sonar.scanner.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:382)
        at org.sonar.scanner.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:351)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
        at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:141)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
        at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
        at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:72)
        at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:66)
        at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
        at com.sun.proxy.$Proxy0.execute(Unknown Source)
        at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
        at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
        at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
        at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
        at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.IllegalArgumentException: 26 is not a valid line offset for pointer. File [...] has 23 character(s) at line 98
        at org.sonar.api.utils.Preconditions.checkArgument(Preconditions.java:43)
        at org.sonar.api.batch.fs.internal.DefaultInputFile.checkValid(DefaultInputFile.java:339)
        at org.sonar.api.batch.fs.internal.DefaultInputFile.newPointer(DefaultInputFile.java:272)
        at org.sonar.api.batch.fs.internal.DefaultInputFile.newRange(DefaultInputFile.java:288)
        at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:95)
        ... 34 more
```

Can you fix this issue ?

--- a/src/main/java/com/github/sbaudoin/sonar/plugins/shellcheck/highlighting/ShellLocation.java
+++ b/src/main/java/com/github/sbaudoin/sonar/plugins/shellcheck/highlighting/ShellLocation.java
@@ -142,7 +142,7 @@ public class ShellLocation {
      * @see #shift(int)
      */
     private ShellLocation shift(char c) {
-        if (c == '\n') {
+        if (c == '\n' || c == '\r') {
             return new ShellLocation(content, line + 1, 1, characterOffset + 1);
         }
         return new ShellLocation(content, line, column + 1, characterOffset + 1);

--- a/src/main/java/com/github/sbaudoin/sonar/plugins/shellcheck/rules/ShellCheckSensor.java
+++ b/src/main/java/com/github/sbaudoin/sonar/plugins/shellcheck/rules/ShellCheckSensor.java
@@ -298,7 +298,7 @@ public class ShellCheckSensor implements Sensor {
                 highlightingData.highlight(highlighting);
             }
             highlighting.save();
-        } catch (IOException e) {
+        } catch (IllegalArgumentException|IOException e) {
             LOGGER.warn("Unable to highlight code for file " + inputFile.filename(), e);
         }
     }

--- a/src/test/java/com/github/sbaudoin/sonar/plugins/shellcheck/highlighting/ShellHighlightingTest.java
+++ b/src/test/java/com/github/sbaudoin/sonar/plugins/shellcheck/highlighting/ShellHighlightingTest.java
@@ -25,11 +25,14 @@ import org.junit.runner.RunWith;
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
 import org.powermock.reflect.Whitebox;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.sensor.highlighting.NewHighlighting;
 import org.sonar.api.batch.sensor.highlighting.TypeOfText;
 import org.sonar.api.utils.log.LogTester;
 import org.sonar.api.utils.log.LoggerLevel;
 
 import java.io.IOException;
+import java.util.List;
 
 import static org.junit.Assert.*;
 import static org.powermock.api.mockito.PowerMockito.*;
@@ -116,6 +119,44 @@ public class ShellHighlightingTest {
         // FI_KEYWORD
         assertHighlightingData(sh.getHighlightingData().get(21), 19, 1, 19, 3, TypeOfText.KEYWORD);
     }
+
+    @Test
+    public void testHighlightLineBreaks() {
+        String[] scripts = new String[] {
+                "#!/bin/bash\n" +
+                        "\n" +
+                        "echo \"Hello world!\"\n",
+                "#!/bin/bash\r" +
+                        "\r" +
+                        "echo \"Hello world!\"\r",
+                "#!/bin/bash\r\n" +
+                        "\r\n" +
+                        "echo \"Hello world!\"\r\n"
+        };
+
+        for (int i = 0; i < scripts.length; i++) {
+            String script = scripts[i];
+            ShellHighlighting sh = new ShellHighlighting(script);
+            List<HighlightingData> data = sh.getHighlightingData();
+            assertEquals(4, data.size());
+            assertHighlightingData(data.get(0), 1, 1, (i < 2)?2:3, 1, TypeOfText.COMMENT);
+            assertHighlightingData(data.get(1), 3, 6, 3, 7, TypeOfText.STRING);
+            assertHighlightingData(data.get(2), 3, 7, 3, 19, TypeOfText.STRING);
+            assertHighlightingData(data.get(3), 3, 19, 3, 20, TypeOfText.STRING);
+        }
+
+        ShellHighlighting sh = new ShellHighlighting("#!/bin/bash\n" +
+                "\n" +
+                "echo \"Hello\rworld!\"\n"
+        );
+        List<HighlightingData> data = sh.getHighlightingData();
+        assertEquals(4, data.size());
+        assertHighlightingData(data.get(0), 1, 1, 2, 1, TypeOfText.COMMENT);
+        assertHighlightingData(data.get(1), 3, 6, 3, 7, TypeOfText.STRING);
+        assertHighlightingData(data.get(2), 3, 7, 4, 7, TypeOfText.STRING);
+        assertHighlightingData(data.get(3), 4, 7, 4, 8, TypeOfText.STRING);
+    }
+
 
     private void assertHighlightingData(HighlightingData hd, int startLine, int startColumnIndex, int endLine, int endColumnIndex, TypeOfText typeOfText) {
         assertEquals(startLine, hd.getStartLine());

